clustering_code/edchan-kmeans.py

```diff
@@ -0,0 +1,913 @@
+                blue_da = rxr.open_rasterio(filepath, masked=True)
+                blue_da.name = 'B02_Blue'
+                blue_das.append(blue_da)
+            elif row.band == 'B03':
+                green_da = rxr.open_rasterio(filepath, masked=True)
+                green_da.name = 'B03_Green'
+                green_das.append(green_da)
+            elif row.band == 'B04':
+                red_da = rxr.open_rasterio(filepath, masked=True)
+                red_da.name = 'B04_Red'
+                red_das.append(red_da)
+            elif row.band == 'B05':
+                nir_da = rxr.open_rasterio(filepath, masked=True)
+                nir_da.name = 'B05_NIR'
+                nir_das.append(nir_da)
+            elif row.band == 'B06':
+                swir1_da = rxr.open_rasterio(filepath, masked=True)
+                swir1_da.name = 'B06_SWIR1'
+                swir1_das.append(swir1_da)
+            elif row.band == 'B07':
+                swir2_da = rxr.open_rasterio(filepath, masked=True)
+                swir2_da.name = 'B07_SWIR2'
+                swir2_das.append(swir2_da)
+            elif row.band == 'Fmask':
+                fmask_da = rxr.open_rasterio(filepath, masked=True)
+                fmask_das.append(fmask_da)
+    merged_blue = rxrmerge.merge_arrays(blue_das)
+    merged_green = rxrmerge.merge_arrays(green_das)
+    merged_red = rxrmerge.merge_arrays(red_das)
+    merged_nir = rxrmerge.merge_arrays(nir_das)
+    merged_swir1 = rxrmerge.merge_arrays(swir1_das)
+    merged_swir2 = rxrmerge.merge_arrays(swir2_das)
+    merged_fmask = rxrmerge.merge_arrays(fmask_das)
+
+    aoi = study_site.to_crs(merged_red.rio.crs)
+
+    clipped_fmask = merged_fmask.rio.clip_box(*aoi.total_bounds)
+    clipped_red = merged_red.rio.clip_box(*aoi.total_bounds)
+    clipped_green = merged_green.rio.clip_box(*aoi.total_bounds)
+    clipped_blue = merged_blue.rio.clip_box(*aoi.total_bounds)
+    clipped_nir = merged_nir.rio.clip_box(*aoi.total_bounds)
+    clipped_swir1 = merged_swir1.rio.clip_box(*aoi.total_bounds)
+    clipped_swir2 = merged_swir2.rio.clip_box(*aoi.total_bounds)
+
+    bits = np.unpackbits(clipped_fmask.values.astype(np.uint8)).reshape(clipped_fmask.shape + (-1,))
+    cloud_mask = np.prod(bits[..., mask_bits] == 0, axis=-1)
+
+    masked_red = clipped_red.where(cloud_mask)
+    masked_green = clipped_green.where(cloud_mask)
+    masked_blue = clipped_blue.where(cloud_mask)
+    masked_nir = clipped_nir.where(cloud_mask)
+    masked_swir1 = clipped_swir1.where(cloud_mask)
+    masked_swir2 = clipped_swir2.where(cloud_mask)
+    processed_granules.append({
+        'date': date,
+        'red': masked_red,
+        'green': masked_green,
+        'blue': masked_blue,
+        'nir': masked_nir,
+        'swir1': masked_swir1,
+        'swir2': masked_swir2
+    })
+
+# %%
+output_folder = 'matplotlib'
+if not os.path.exists(output_folder):
+    os.makedirs(output_folder)
+for processed_granule in processed_granules:
```

Comment:
Here you could write a function to squeeze, convert nan to num, and then concat bands into a single data array for each date, and assign a date dimension. This could make the code shorter, more modular, and easier to read.